### ISSUE TITLE
Warnings logged for null values

### ISSUE THREAD

**Issue by hc4** [Graylog2/graylog-plugin-pipeline-processor#63]
### Problem description

If you pass **null** to **to_string** (`to_string(m["3"],"")`), then this line will be logged:
`WARN  [EqualityExpression] left expression evaluated to null, returning false: to_string(default: "", value: m["3"])`

If you try to compare **null** to something (`m["3"]!=""`), it will log:
`WARN  [EqualityExpression] left expression evaluated to null, returning false: m["3"]`

### Environment
* Pipeline Processor plugin version: 1.1.0-beta1

**Comment by joschi** [comment]
Those should definitely be logged on DEBUG level.

**Comment by kroepke** [comment]
I tend to disagree but the situation is a little more involved, let me explain.

The language is designed to make mistakes very painful, largely because it is intended to be a very low level tool for which we want to write code generators in the future.
Thus the early decision was to make subtle failures like `null` comparisons/usage loud because `null` is mostly an error in writing the rule.

I realize that in this case the result is a regex matcher group map, in which certain groups are optional.
Personally I'd rather improve the handling around those than silencing `null` warnings. My reason is that whenever the interpreter encounters a `null` it usually cannot continue meaningfully. It also doesn't know the outside context, such as the expression being ok to be `null`.

Another underlying problem is that we don't currently have a good way to record errors to aid in debugging rules when problems like this pop up. The challenge here is that it needs to be so fast as to not impact processing speed too much. I'm still looking for a solution to that.

**Comment by hc4** [comment]
What's the meaning of default value for **to_string**, if it won't accept **null**?
Or it will accept, but also will log WARN?
If thing there is situations, when **null** value is expected to be fin.
I think behaviour should be like with NullReferenceException - it should abort processing if **null** value received in unexpected place.

**Comment by kroepke** [comment]
Just had a look at the code, the `null` handling in `to_string` seems to have a bug.
The intention is to use the default value if the required argument was `null` but it returns `null` directly.
The warning you are seeing is from later up the evaluation chain, and not from the function itself.
I'm adding a test case to verify.

**Comment by kroepke** [comment]
Yup, it's a bug.
Many thanks for your contributions and feedback, it's very valuable and much appreciated!
